### PR TITLE
Add streaming benchmarks

### PR DESCRIPTION
**Pull request by JamesNK**
https://github.com/aspnet/SignalR/issues/1811

### CODE REVIEW

**Comment by JamesNK** [comment]
Results are odd. For some reason streaming zero results is very slow.

```
                            Method |       Mean |      Error |     StdDev |      Op/s |  Gen 0 |  Gen 1 | Allocated |
---------------------------------- |-----------:|-----------:|-----------:|----------:|-------:|-------:|----------:|
                        Invocation |   1.392 us |  0.1497 us |  0.1664 us | 718,433.8 | 0.0076 |      - |     704 B |
                   InvocationAsync |   1.409 us |  0.1275 us |  0.1365 us | 709,597.8 | 0.0076 |      - |     704 B |
             InvocationReturnValue |   1.522 us |  0.2632 us |  0.2926 us | 657,232.2 | 0.0076 |      - |     800 B |
             InvocationReturnAsync |   1.398 us |  0.0213 us |  0.0219 us | 715,554.6 | 0.0095 |      - |     896 B |
          InvocationValueTaskAsync |   1.824 us |  0.2017 us |  0.2159 us | 548,207.1 | 0.0095 |      - |     864 B |
                  StreamObservable |  20.557 us |  0.8067 us |  0.7546 us |  48,645.4 |      - |      - |    5346 B |
             StreamObservableAsync |  16.104 us |  2.1547 us |  2.3055 us |  62,096.2 |      - |      - |    5442 B |
    StreamObservableValueTaskAsync |  17.713 us |  0.6642 us |  0.7649 us |  56,456.4 | 0.0153 |      - |    5482 B |
               StreamChannelReader |   6.225 us |  0.3830 us |  0.4410 us | 160,638.3 | 0.0076 |      - |    3544 B |
          StreamChannelReaderAsync |   6.630 us |  0.4299 us |  0.4951 us | 150,836.2 | 0.0076 |      - |    3640 B |
 StreamChannelReaderValueTaskAsync |   6.380 us |  0.3970 us |  0.4572 us | 156,738.9 | 0.0076 |      - |    3624 B |
     StreamChannelReaderCount_Zero |  78.498 us |  3.4562 us |  3.9801 us |  12,739.1 |      - |      - |    4251 B |
      StreamChannelReaderCount_Ten |  18.066 us |  0.8389 us |  0.9660 us |  55,353.4 | 0.0305 |      - |    4789 B |
 StreamChannelReaderCount_Thousand | 215.416 us | 10.4302 us | 12.0114 us |   4,642.2 | 2.4414 | 0.4883 |    7816 B |
```

File: benchmarks/Microsoft.AspNetCore.SignalR.Microbenchmarks/DefaultHubDispatcherBenchmark.cs [hunk 1]
```diff
@@ -157,6 +157,22 @@ public ValueTask<ChannelReader<int>> StreamChannelReaderValueTaskAsync()
 
                 return new ValueTask<ChannelReader<int>>(channel);
             }
+
+            public ChannelReader<string> StreamChannelReaderCount(int count)
```
Comment (davidfowl) [inline benchmarks/Microsoft.AspNetCore.SignalR.Microbenchmarks/DefaultHubDispatcherBenchmark.cs]: Add an observable as well, don't convert to string, just return the ints.

**Comment by JamesNK** [comment]
StreamChannelReaderCount_Zero  is weird.

```
                            Method |       Mean |     Error |    StdDev |      Op/s |   Gen 0 |  Gen 1 |  Gen 2 | Allocated |
---------------------------------- |-----------:|----------:|----------:|----------:|--------:|-------:|-------:|----------:|
                        Invocation |   1.649 us | 0.0099 us | 0.0088 us | 606,548.4 |  0.0782 |      - |      - |     704 B |
                   InvocationAsync |   1.687 us | 0.0382 us | 0.0319 us | 592,904.1 |  0.0782 |      - |      - |     704 B |
             InvocationReturnValue |   1.703 us | 0.0102 us | 0.0095 us | 587,128.2 |  0.0877 |      - |      - |     800 B |
             InvocationReturnAsync |   1.805 us | 0.0152 us | 0.0142 us | 554,164.1 |  0.0992 |      - |      - |     896 B |
          InvocationValueTaskAsync |   1.848 us | 0.0439 us | 0.0389 us | 541,068.2 |  0.0954 |      - |      - |     864 B |
                  StreamObservable |  14.937 us | 0.8740 us | 0.8584 us |  66,947.4 |  0.0153 |      - |      - |    5530 B |
             StreamObservableAsync |  15.684 us | 0.7695 us | 0.7557 us |  63,760.2 |  0.0153 |      - |      - |    5649 B |
    StreamObservableValueTaskAsync |  15.972 us | 0.8684 us | 0.8528 us |  62,608.2 |  0.0153 |      - |      - |    5633 B |
               StreamChannelReader |   5.797 us | 0.2854 us | 0.2803 us | 172,504.2 |  0.0076 |      - |      - |    3608 B |
          StreamChannelReaderAsync |   6.243 us | 0.2191 us | 0.2152 us | 160,169.7 |  0.0076 |      - |      - |    3704 B |
 StreamChannelReaderValueTaskAsync |   6.005 us | 0.0441 us | 0.0391 us | 166,538.4 |  0.0076 |      - |      - |    3688 B |
     StreamChannelReaderCount_Zero |  67.483 us | 3.5596 us | 3.1555 us |  14,818.5 |  0.1221 |      - |      - |    4174 B |
      StreamChannelReaderCount_One |   8.661 us | 0.5244 us | 0.4905 us | 115,460.9 |  0.0153 |      - |      - |    4005 B |
 StreamChannelReaderCount_Thousand | 155.379 us | 2.7378 us | 2.6888 us |   6,435.9 | 22.4609 | 0.2441 |      - |   17058 B |
        StreamObservableCount_Zero |  15.408 us | 0.6304 us | 0.5896 us |  64,902.1 |  0.0153 |      - |      - |    5753 B |
         StreamObservableCount_One | 142.925 us | 3.3967 us | 3.3360 us |   6,996.7 |  2.1973 | 0.4883 | 0.2441 |    6450 B |
    StreamObservableCount_Thousand | 379.627 us | 4.4100 us | 4.3312 us |   2,634.2 | 28.3203 | 0.4883 |      - |    8516 B |
```

File: benchmarks/Microsoft.AspNetCore.SignalR.Microbenchmarks/DefaultHubDispatcherBenchmark.cs [hunk 2]
```diff
@@ -157,6 +157,29 @@ public ValueTask<ChannelReader<int>> StreamChannelReaderValueTaskAsync()
...
+                    }
+                    channel.Writer.Complete();
+                });
+
+                return channel.Reader;
+            }
+
+            public IObservable<int> StreamObservableCount(int count)
+            {
+                return Observable.Interval(TimeSpan.Zero)
```
Comment (davidfowl) [inline benchmarks/Microsoft.AspNetCore.SignalR.Microbenchmarks/DefaultHubDispatcherBenchmark.cs]: Is this the right way to do this?

**Comment by BrennanConroy** [comment]
>StreamChannelReaderCount_Zero is weirdly slow.
StreamObservableCount_Thousand uses a weirdly large amount of memory.

Any ideas why?

**Comment by JamesNK** [comment]
No. Should the issue be reopened to investigate/solve?

**Pull request merged**